Source/Libraries/GSF.Security/OIDCSecurityProvider.cs

```diff
@@ -0,0 +1,619 @@
+            using (HttpRequestMessage request = new HttpRequestMessage())
+            {
+                ConfigureRequest(request);
+                using (HttpClient client = new HttpClient())
+                {
+                    response = client.SendAsync(request).Result;
+                    if (!response.IsSuccessStatusCode)
+                        return null;
+                    r = JsonConvert.DeserializeObject<TokenResponse>(response.Content.ReadAsStringAsync().Result);
+                }
+            }
+
+            return r;
+            
+        }
+
+        private bool DecodeToken(TokenResponse token)
+        {
+
+            try
+            {
+                IJsonSerializer serializer = new JsonNetSerializer();
+                var provider = new UtcDateTimeProvider();
+                IJwtValidator validator = new JwtValidator(serializer, provider);
+                IBase64UrlEncoder urlEncoder = new JwtBase64UrlEncoder();
+                IJwtAlgorithm algorithm = new HMACSHA256Algorithm(); // symmetric
+                IJwtDecoder decoder = new JwtDecoder(serializer, validator, urlEncoder, algorithm);
+
+                // we do not actually validate the Signature of the Token
+                // That is necessary to allow self-signed Tokens from the openXDA
+                IDictionary<string, object> tokenContent = decoder.DecodeToObject<IDictionary<string, object>>(token.id_token , m_ClientSecret, verify: false);
+            
+                // Translate UserDetails according to Token
+                UserData userData = new UserData(tokenContent.GetOrDefault("sub").ToString());
+
+                // Initialize user data.
+                userData.Initialize();
+
+                userData.Username = tokenContent.GetOrDefault("name").ToString();
+                
+                if (tokenContent.ContainsKey("given_name"))
+                    userData.FirstName = tokenContent.GetOrDefault("given_name").ToString();
+
+                if (tokenContent.ContainsKey("family_name"))
+                    userData.LastName = tokenContent.GetOrDefault("family_name").ToString();
+
+                if (tokenContent.ContainsKey("phone_number"))
+                    userData.PhoneNumber = tokenContent.GetOrDefault("phone_number").ToString();
+
+                if (tokenContent.ContainsKey("email"))
+                    userData.EmailAddress = tokenContent.GetOrDefault("email").ToString();
+
+                try
+                {
+                    if (tokenContent.ContainsKey(m_RolesClaim))
+                        userData.Roles = (List<string>)(tokenContent.GetOrDefault(m_RolesClaim));
+                }
+                catch (Exception ex)
+                {
+                    LastException = ex;
+                    Log.Publish(MessageLevel.Warning, MessageFlags.SecurityMessage, "DecodingTokenError", "Failed to decode Roles Claim.", "Failed to decode Claim for roles.", ex);
+                }
```

Comment:
`tokenContent.GetOrDefault("...").ToString()` looks like a recipe for a `NullReferenceException`.

If the field is expected to be there, consider using the indexer `tokenContent["..."].ToString()` instead. Even better would be to use `TryGetValue()` and throw an appropriate exception.

```c#
if (!tokenContent.TryGetValue("sub", out object sub))
    throw new Exception("Token data was not derived from valid JWT: missing 'sub'");

UserData userData = new UserData(sub.ToString());
```

For optional parameters, you can just flip the condition on the if statement.

```c#
if (tokenContent.TryGetValue("given_name", out object firstName))
    userData.FirstName = firstName.ToString();
```